### PR TITLE
KYLIN-4117 Auto adjust data type of RelNode for intersect_count

### PR DESCRIPTION
**Pull request by hit-lacus**
### Issue description

In query of intersect_count, says `intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01',date '2012-01-02'])`, we need to specific data type of each entry in array, if *date* miss, return value will be 0, which will mislead user. So I think we should auto adjust data type of each entry from string/char to real data type.

image

### CODE REVIEW

**Comment by hit-lacus** [comment]
### Issue description


### Root Cause Analysis

### Desgin Evidence

### Test Evidence

```sql
select 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-03']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02', '2012-01-03'])
from KYLIN_SALES 
where part_dt in ('2012-01-01', '2012-01-02', '2012-01-03')


select 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-03']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01',date '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01',date '2012-01-02',date '2012-01-03'])
from KYLIN_SALES 
where part_dt in ('2012-01-01', '2012-01-02', '2012-01-03')


select  LSTG_FORMAT_NAME, 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-03']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02', '2012-01-03'])
from KYLIN_SALES 
where part_dt in ('2012-01-01', '2012-01-02', '2012-01-03')
group by LSTG_FORMAT_NAME


select  
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 15]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3, 15])
from KYLIN_SALES 
where lstg_site_id in (0, 3, 15)


select LSTG_FORMAT_NAME, 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 15]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3, 15])
from KYLIN_SALES 
group by LSTG_FORMAT_NAME
where lstg_site_id in (0, 3, 15)

```

**Comment by hit-lacus** [comment]
### Desgin Evidence

**Comment by hit-lacus** [comment]
### Test Evidence

```sql
select 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-03']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02', '2012-01-03'])
from KYLIN_SALES 
where part_dt in ('2012-01-01', '2012-01-02', '2012-01-03')


select 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-03']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01',date '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[date '2012-01-01',date '2012-01-02',date '2012-01-03'])
from KYLIN_SALES 
where part_dt in ('2012-01-01', '2012-01-02', '2012-01-03')


select  LSTG_FORMAT_NAME, 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-03']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02']), 
intersect_count(OPS_USER_ID, part_dt, array[ '2012-01-01', '2012-01-02', '2012-01-03'])
from KYLIN_SALES 
where part_dt in ('2012-01-01', '2012-01-02', '2012-01-03')
group by LSTG_FORMAT_NAME


select  
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 15]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3, 15])
from KYLIN_SALES 
where lstg_site_id in (0, 3, 15)


select LSTG_FORMAT_NAME, 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 15]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3]), 
intersect_count(OPS_USER_ID, lstg_site_id, array[ 0, 3, 15])
from KYLIN_SALES 
group by LSTG_FORMAT_NAME
where lstg_site_id in (0, 3, 15)

```

**Comment by nichunen** [comment]
@hit-lacus Would you please modify the test sql in IT?

File: query/src/main/java/org/apache/kylin/query/relnode/OLAPProjectRel.java [hunk 1]
```diff
@@ -137,6 +158,16 @@ public void implementOLAP(OLAPImplementor implementor) {
         this.afterAggregate = context.afterAggregate;
 
         this.columnRowType = buildColumnRowType();
+        RelNode parentNode = implementor.getParentNode();
+        if (parentNode instanceof OLAPAggregateRel) {
+            OLAPAggregateRel rel = (OLAPAggregateRel) parentNode;
+            for (AggregateCall call : rel.getRewriteAggCalls()) {
+                if (call.getAggregation().getName().equalsIgnoreCase(BitmapMeasureType.FUNC_INTERSECT_COUNT_DISTINCT)) {
+                    hasIntersect = true;
+                    logger.trace("Find intersect in query.");
```
[the remark this answers is not in the document]
Comment (hit-lacus) [reply query/src/main/java/org/apache/kylin/query/relnode/OLAPProjectRel.java]: OK, I got it.

**Comment by hit-lacus** [comment]
Add test for timestamp data type.

```sql
select intersect_count(city, log_time, ARRAY['2019-01-01 11:32:10', '2019-01-01 11:32:08'])
from USERACTIONLOG

select intersect_count(city, register_date, ARRAY['2017-01-01', '2017-01-02'])
from USERACTIONLOG

select intersect_count(city, page_id, ARRAY[1, 2])
from USERACTIONLOG

```

**Pull request merged**